Add API, actions, and reducers to get Triggers resources

a-roberts opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

For https://github.com/tektoncd/dashboard/issues/665

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

Hold on, think this is the code but let's see if it actually gets ok

Manual test:

```
adams-mbp:tektoncd aroberts$ k get el
NAME                   AGE
listener               24s
listener-interceptor   84s
adams-mbp:tektoncd aroberts$ k get tt
NAME                AGE
pipeline-template   100s
adams-mbp:tektoncd aroberts$ k get tb
NAME               AGE
pipeline-binding   104s
```

(from Triggers example repo), then using the API gives...


http://localhost:9097/proxy/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/triggertemplates/
->
```
{"apiVersion":"tekton.dev/v1alpha1","items":[{"apiVersion":"tekton.dev/v1alpha1","kind":"TriggerTemplate","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"tekton.dev/v1alpha1\",\"kind\":\"TriggerTemplate\",\"metadata\":{\"annotations\":{},\"name\":\"pipeline-template\",\"namespace\":\"tekton-pipelines\"},\"spec\":{\"params\":[{\"default\":\"master\",\"description\":\"The git revision\",\"name\":\"gitrevision\"},{\"description\":\"The git repository url\",\"name\":\"gitrepositoryurl\"},{\"default\":\"This is the default message\",\"description\":\"The message to print\",\"name\":\"message\"},{\"description\":\"The Content-Type of the event\",\"name\":\"contenttype\"}],\"resourcetemplates\":[{\"apiVersion\":\"tekton.dev/v1alpha1\",\"kind\":\"PipelineResource\",\"metadata\":{\"name\":\"git-source-$(uid)\"},\"spec\":{\"params\":[{\"name\":\"revision\",\"value\":\"$(params.gitrevision)\"},{\"name\":\"url\",\"value\":\"$(params.gitrepositoryurl)\"}],\"type\":\"git\"}},{\"apiVersion\":\"tekton.dev/v1alpha1\",\"kind\":\"PipelineRun\",\"metadata\":{\"generateName\":\"simple-pipeline-run\"},\"spec\":{\"params\":[{\"name\":\"message\",\"value\":\"$(params.message)\"},{\"name\":\"contenttype\",\"value\":\"$(params.contenttype)\"}],\"pipelineRef\":{\"name\":\"simple-pipeline\"},\"resources\":[{\"name\":\"git-source\",\"resourceRef\":{\"name\":\"git-source-$(uid)\"}}]}}]}}\n"},"creationTimestamp":"2019-11-01T11:33:31Z","generation":1,"name":"pipeline-template","namespace":"tekton-pipelines","resourceVersion":"7323","selfLink":"/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/triggertemplates/pipeline-template","uid":"6e7b302a-fc9b-11e9-a4cd-025000000001"},"spec":{"params":[{"default":"master","description":"The git revision","name":"gitrevision"},{"description":"The git repository url","name":"gitrepositoryurl"},{"default":"This is the default message","description":"The message to print","name":"message"},{"description":"The Content-Type of the event","name":"contenttype"}],"resourcetemplates":[{"apiVersion":"tekton.dev/v1alpha1","kind":"PipelineResource","metadata":{"name":"git-source-$(uid)"},"spec":{"params":[{"name":"revision","value":"$(params.gitrevision)"},{"name":"url","value":"$(params.gitrepositoryurl)"}],"type":"git"}},{"apiVersion":"tekton.dev/v1alpha1","kind":"PipelineRun","metadata":{"generateName":"simple-pipeline-run"},"spec":{"params":[{"name":"message","value":"$(params.message)"},{"name":"contenttype","value":"$(params.contenttype)"}],"pipelineRef":{"name":"simple-pipeline"},"resources":[{"name":"git-source","resourceRef":{"name":"git-source-$(uid)"}}]}}]}}],"kind":"TriggerTemplateList","metadata":{"continue":"","resourceVersion":"8348","selfLink":"/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/triggertemplates"}}
```

http://localhost:9097/proxy/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/triggerbindings/
->
```
{"apiVersion":"tekton.dev/v1alpha1","items":[{"apiVersion":"tekton.dev/v1alpha1","kind":"TriggerBinding","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"tekton.dev/v1alpha1\",\"kind\":\"TriggerBinding\",\"metadata\":{\"annotations\":{},\"name\":\"pipeline-binding\",\"namespace\":\"tekton-pipelines\"},\"spec\":{\"params\":[{\"name\":\"gitrevision\",\"value\":\"$(body.head_commit.id)\"},{\"name\":\"gitrepositoryurl\",\"value\":\"$(body.repository.url)\"},{\"name\":\"contenttype\",\"value\":\"$(header.Content-Type)\"}]}}\n"},"creationTimestamp":"2019-11-01T11:33:28Z","generation":1,"name":"pipeline-binding","namespace":"tekton-pipelines","resourceVersion":"7319","selfLink":"/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/triggerbindings/pipeline-binding","uid":"6c66696f-fc9b-11e9-a4cd-025000000001"},"spec":{"params":[{"name":"gitrevision","value":"$(body.head_commit.id)"},{"name":"gitrepositoryurl","value":"$(body.repository.url)"},{"name":"contenttype","value":"$(header.Content-Type)"}]}}],"kind":"TriggerBindingList","metadata":{"continue":"","resourceVersion":"8355","selfLink":"/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/triggerbindings"}}
```

http://localhost:9097/proxy/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/eventlisteners/
->
```
{"apiVersion":"tekton.dev/v1alpha1","items":[{"apiVersion":"tekton.dev/v1alpha1","kind":"EventListener","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"tekton.dev/v1alpha1\",\"kind\":\"EventListener\",\"metadata\":{\"annotations\":{},\"name\":\"listener\",\"namespace\":\"tekton-pipelines\"},\"spec\":{\"serviceAccountName\":\"tekton-webhooks-extension\",\"triggers\":[{\"binding\":{\"name\":\"pipeline-binding\"},\"interceptor\":{\"objectRef\":{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"name\":\"el-listener-interceptor\",\"namespace\":\"tekton-pipelines\"}},\"name\":\"foo-trig\",\"params\":[{\"name\":\"secretKeyRef\",\"value\":\"secretToken\"},{\"name\":\"wantedRepositoryURL\",\"value\":\"https://github.ibm.com/AROBERTS/devops-world-example.git\"}],\"template\":{\"name\":\"pipeline-template\"}}]}}\n"},"creationTimestamp":"2019-11-01T11:34:45Z","generation":1,"name":"listener","namespace":"tekton-pipelines","resourceVersion":"8297","selfLink":"/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/eventlisteners/listener","uid":"9a7b9378-fc9b-11e9-a4cd-025000000001"},"spec":{"serviceAccountName":"tekton-webhooks-extension","triggers":[{"binding":{"name":"pipeline-binding"},"interceptor":{"objectRef":{"apiVersion":"v1","kind":"Service","name":"el-listener-interceptor","namespace":"tekton-pipelines"}},"name":"foo-trig","params":[{"name":"secretKeyRef","value":"secretToken"},{"name":"wantedRepositoryURL","value":"https://github.ibm.com/AROBERTS/devops-world-example.git"}],"template":{"name":"pipeline-template"}}]},"status":{"address":{"hostname":"el-listener.tekton-pipelines.svc.cluster.local"},"conditions":[{"lastTransitionTime":"2019-11-01T11:34:45Z","message":"Deployment does not have minimum availability.","reason":"MinimumReplicasUnavailable","status":"False","type":"Available"},{"lastTransitionTime":"2019-11-01T11:34:45Z","message":"Deployment exists","status":"True","type":"Deployment"},{"lastTransitionTime":"2019-11-01T11:44:45Z","message":"ReplicaSet \"el-listener-b486cb988\" has timed out progressing.","reason":"ProgressDeadlineExceeded","status":"False","type":"Progressing"},{"lastTransitionTime":"2019-11-01T11:44:45Z","message":"pods \"el-listener-b486cb988-\" is forbidden: error looking up service account tekton-pipelines/tekton-webhooks-extension: serviceaccount \"tekton-webhooks-extension\" not found","reason":"FailedCreate","status":"True","type":"ReplicaFailure"},{"lastTransitionTime":"2019-11-01T11:34:45Z","message":"Service exists","status":"True","type":"Service"}],"configuration":{"generatedName":"el-listener"}}},{"apiVersion":"tekton.dev/v1alpha1","kind":"EventListener","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"tekton.dev/v1alpha1\",\"kind\":\"EventListener\",\"metadata\":{\"annotations\":{},\"name\":\"listener-interceptor\",\"namespace\":\"tekton-pipelines\"},\"spec\":{\"serviceAccountName\":\"tekton-triggers-example-sa\",\"triggers\":[{\"binding\":{\"name\":\"pipeline-binding\"},\"interceptor\":{\"header\":[{\"name\":\"Foo-Trig-Header1\",\"value\":\"string-value\"},{\"name\":\"Foo-Trig-Header2\",\"value\":[\"array-val1\",\"array-val2\"]}],\"objectRef\":{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"name\":\"gh-validate\",\"namespace\":\"default\"}},\"name\":\"foo-trig\",\"params\":[{\"name\":\"message\",\"value\":\"Hello from the Triggers EventListener!\"}],\"template\":{\"name\":\"pipeline-template\"}}]}}\n"},"creationTimestamp":"2019-11-01T11:33:45Z","generation":1,"name":"listener-interceptor","namespace":"tekton-pipelines","resourceVersion":"8221","selfLink":"/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/eventlisteners/listener-interceptor","uid":"770d6d75-fc9b-11e9-a4cd-025000000001"},"spec":{"serviceAccountName":"tekton-triggers-example-sa","triggers":[{"binding":{"name":"pipeline-binding"},"interceptor":{"header":[{"name":"Foo-Trig-Header1","value":"string-value"},{"name":"Foo-Trig-Header2","value":["array-val1","array-val2"]}],"objectRef":{"apiVersion":"v1","kind":"Service","name":"gh-validate","namespace":"default"}},"name":"foo-trig","params":[{"name":"message","value":"Hello from the Triggers EventListener!"}],"template":{"name":"pipeline-template"}}]},"status":{"address":{"hostname":"el-listener-interceptor.tekton-pipelines.svc.cluster.local"},"conditions":[{"lastTransitionTime":"2019-11-01T11:33:45Z","message":"Deployment does not have minimum availability.","reason":"MinimumReplicasUnavailable","status":"False","type":"Available"},{"lastTransitionTime":"2019-11-01T11:33:45Z","message":"Deployment exists","status":"True","type":"Deployment"},{"lastTransitionTime":"2019-11-01T11:43:45Z","message":"ReplicaSet \"el-listener-interceptor-79f7976f66\" has timed out progressing.","reason":"ProgressDeadlineExceeded","status":"False","type":"Progressing"},{"lastTransitionTime":"2019-11-01T11:43:45Z","message":"pods \"el-listener-interceptor-79f7976f66-\" is forbidden: error looking up service account tekton-pipelines/tekton-triggers-example-sa: serviceaccount \"tekton-triggers-example-sa\" not found","reason":"FailedCreate","status":"True","type":"ReplicaFailure"},{"lastTransitionTime":"2019-11-01T11:33:45Z","message":"Service exists","status":"True","type":"Service"}],"configuration":{"generatedName":"el-listener-interceptor"}}}],"kind":"EventListenerList","metadata":{"continue":"","resourceVersion":"8357","selfLink":"/apis/tekton.dev/v1alpha1/namespaces/tekton-pipelines/eventlisteners"}}
```

a-roberts commented:
/rest

```
I1101 11:46:10.351] --- FAIL: TestExtensionRegistration (10.04s)
I1101 11:46:10.351]     routes_test.go:318: In TestExtensionRegistration
I1101 11:46:10.351]     routes_test.go:344: robots.secrets
I1101 11:46:10.351]     routes_test.go:461: waiting for registration
I1101 11:46:10.352]     routes_test.go:462: services:
I1101 11:46:10.352]     routes_test.go:463: start time: 2019-11-01 11:46:00.236459867 +0000 UTC m=+0.188153595
I1101 11:46:10.352]     routes_test.go:464: --- ---
... skipping 5 lines ...
I1101 11:46:10.353]     routes_test.go:478: counter: 2
I1101 11:46:10.353]     routes_test.go:479: time: 2019-11-01 11:46:00.247983598 +0000 UTC m=+0.199677396
I1101 11:46:10.353]     routes_test.go:483: extensionCreates:
I1101 11:46:10.353]     routes_test.go:484: 2
I1101 11:46:10.353]     routes_test.go:490: 
I1101 11:46:10.353]     routes_test.go:470: Timed out waiting for expected services to be registered
I1101 11:46:10.354] FAIL
I1101 11:46:10.354] FAIL	github.com/tektoncd/dashboard/pkg/router	10.219s
I1101 11:46:10.354] ?   	github.com/tektoncd/dashboard/pkg/testutils	[no test files]
I1101 11:46:10.354] ?   	github.com/tektoncd/dashboard/pkg/utils	[no test files]
I1101 11:46:10.354] ?   	github.com/tektoncd/dashboard/pkg/websocket	[no test files]
```

🙄

a-roberts commented:
/retest

a-roberts commented:
Plumbing chinwag raised: https://tektoncd.slack.com/archives/CJ4ERJWAU/p1572610739084900

a-roberts commented:
/retest

carlos-logro commented on src/api/index.js:
```diff
@@ -404,3 +404,33 @@ export async function determineInstallNamespace() {
+  return get(uri);
+}
+
+export function getTriggerBindings({ namespace } = {}) {
```
shouldn't there be `filters` argument as well?

a-roberts replied:
Added, yeah I'll add those next thanks

a-roberts commented:
/test pull-tekton-dashboard-integration-tests

a-roberts commented:
https://github.com/tektoncd/dashboard/issues/447 issue made for dealing with the failing integration tests

ncskier commented on src/actions/secrets.js:
```diff
@@ -78,7 +78,10 @@ export function deleteSecret(secrets) {
+    namespacesToSecretsMap.forEach(async function handleSecretUnpatching(
+      value,
+      namespace
+    ) {
```
Was this an intentional change to have in this PR? Seems unrelated to the rest of the PR (not that this is a bad code change).

a-roberts replied:
Renamed a variable, ran the lint fix thus it's split over the new lines

ncskier commented on src/actions/triggerBindings.test.js:
```diff
@@ -0,0 +1,49 @@
+  expect(creators.fetchNamespacedCollection).toHaveBeenCalledWith(
+    'TriggerBinding',
+    API.getTriggerBindings,
+    { namespace }
```
nit: Please include `filters` in the test (this will be consistent with [`pipelineRuns.test`](https://github.com/tektoncd/dashboard/blob/master/src/actions/pipelineRuns.test.js#L33-L38).

a-roberts replied:
Done

ncskier commented on src/actions/eventListeners.js:
```diff
@@ -0,0 +1,31 @@
+export function fetchEventListeners({ namespace } = {}) {
+  return fetchNamespacedCollection('EventListener', getEventListeners, {
+    namespace
+  });
```
Should this include `filters` in the parameter object to be consistent with `triggerBindings.js` and the `getEventListeners()` function?

If so, we should also add `filters` to the test (see following comment about this for `triggerBindings.test.js`).

a-roberts replied:
Done and done

ncskier commented on src/actions/triggerTemplates.js:
```diff
@@ -0,0 +1,31 @@
+  return fetchNamespacedCollection('TriggerTemplate', getTriggerTemplates, {
+    namespace
+  });
+}
```
Should this include `filters` in the parameter object to be consistent with `triggerBindings.js` and the `getTriggerTemplates()` function?

If so, we should also add `filters` to the test (see following comment about this for `triggerBindings.test.js`).

a-roberts replied:
Yep done and done

ncskier commented on src/api/index.js:
```diff
@@ -404,3 +404,48 @@ export async function determineInstallNamespace() {
+  let queryParams;
+  if (filters.length) {
+    queryParams = { labelSelector: filters };
+  }
```
Since we have this same code snippet copied and pasted for all our `get` functions that use `filters`, should we create a helper function to do this? 😅

a-roberts replied:
```
function getQueryParams(filters) {
  if (filters.length) {
    return { labelSelector: filters };
  }
  return "";
}
```

adding

ncskier commented on src/api/index.test.js:
```diff
@@ -644,3 +644,66 @@ it('getResource', () => {
+    expect(triggerTemplate).toEqual(data);
+    fetchMock.restore();
+  });
+});
```
super nit: you don't have to fix this, but the ordering of the TriggerTemplate stanzas is different from the others (the others are singular then plural) 😁

a-roberts replied:
True, done

ncskier commented on src/reducers/eventListeners.test.js:
```diff
@@ -0,0 +1,59 @@
+import * as selectorCreators from './selectorCreators';
+import createReducer, * as selectors from './eventListeners';
+
+const name = 'pipeline resource name';
```
copy & paste nit: should be `event listener name` 😄

a-roberts replied:
Done 😆

ncskier commented on src/reducers/triggerBindings.test.js:
```diff
@@ -0,0 +1,59 @@
+import * as selectorCreators from './selectorCreators';
+import createReducer, * as selectors from './triggerBindings';
+
+const name = 'pipeline resource name';
```
same copy & paste nit as the `eventListeners.test.js`

a-roberts replied:
😆 fixed

ncskier commented on src/reducers/triggerTemplates.test.js:
```diff
@@ -0,0 +1,61 @@
+import * as selectorCreators from './selectorCreators';
+import createReducer, * as selectors from './triggerTemplates';
+
+const name = 'pipeline resource name';
```
same copy & paste nit as the `eventListeners.test.js`

a-roberts replied:
Fixed

a-roberts commented:
Cheers for the detailed review @ncskier, I'll address these once I fix our build break https://github.com/tektoncd/dashboard/issues/447 then rebase 😄

a-roberts commented:
Rebased

ncskier commented on src/actions/eventListeners.test.js:
```diff
@@ -0,0 +1,49 @@
+  expect(creators.fetchNamespacedCollection).toHaveBeenCalledWith(
+    'EventListener',
+    API.getEventListeners,
+    { namespace }
```
nit: please add the `filters` param to this test (like we have with the TriggerBinding & TriggerTemplate)
```
  const filters = ['someFilter'];
  fetchTriggerBindings({ filters, namespace });
...
    { namespace, filters }
```

a-roberts commented:
Pushed 😄

a-roberts commented:
/approve

carlos-logro commented:
Just noticed that the reducers are not specified in `reducers/index`. Must be added for the panels to work properly.

a-roberts commented:
> Just noticed that the reducers are not specified in `reducers/index`. Must be added for the panels to work properly.

How about this?

```
/*
Copyright 2019 The Tekton Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { combineReducers } from 'redux';

import clusterTasks, * as clusterTaskSelectors from './clusterTasks';
import extensions, * as extensionSelectors from './extensions';
import locale, * as localeSelectors from './locale';
import namespaces, * as namespaceSelectors from './namespaces';
import notifications, * as notificationSelectors from './notifications';
import pipelines, * as pipelineSelectors from './pipelines';
import pipelineResources, * as pipelineResourcesSelectors from './pipelineResources';
import pipelineRuns, * as pipelineRunsSelectors from './pipelineRuns';
import secrets, * as secretSelectors from './secrets';
import serviceAccounts, * as serviceAccountSelectors from './serviceAccounts';
import tasks, * as taskSelectors from './tasks';
import taskRuns, * as taskRunsSelectors from './taskRuns';
import eventListeners, * as eventListenerSelectors from './eventListeners';
import triggerTemplates, * as triggerTemplateSelectors from './triggerTemplates';
import triggerBindings, * as triggerBindingSelectors from './triggerBindings';


export default combineReducers({
  clusterTasks,
  extensions,
  locale,
  namespaces,
  notifications,
  pipelines: pipelines(),
  pipelineResources: pipelineResources(),
  pipelineRuns: pipelineRuns(),
  secrets,
  serviceAccounts: serviceAccounts(),
  tasks: tasks(),
  taskRuns: taskRuns(),
  triggerTemplates: triggerTemplates(),
  triggerBindings: triggerBindings(),
  eventListeners: eventListeners()
});

export function getSelectedNamespace(state) {
  return namespaceSelectors.getSelectedNamespace(state.namespaces);
}

export function getNamespaces(state) {
  return namespaceSelectors.getNamespaces(state.namespaces);
}

export function getServiceAccounts(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return serviceAccountSelectors.getServiceAccounts(
    state.serviceAccounts,
    namespace
  );
}

export function isFetchingServiceAccounts(state) {
  return serviceAccountSelectors.isFetchingServiceAccounts(
    state.serviceAccounts
  );
}

export function getServiceAccountsErrorMessage(state) {
  return serviceAccountSelectors.getServiceAccountsErrorMessage(
    state.serviceAccounts
  );
}

export function isFetchingNamespaces(state) {
  return namespaceSelectors.isFetchingNamespaces(state.namespaces);
}

export function getExtensions(state) {
  return extensionSelectors
    .getExtensions(state.extensions)
    .filter(({ displayName }) => !!displayName);
}

export function getExtensionsErrorMessage(state) {
  return extensionSelectors.getExtensionsErrorMessage(state.extensions);
}

export function isFetchingExtensions(state) {
  return extensionSelectors.isFetchingExtensions(state.extensions);
}

export function getPipeline(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return pipelineSelectors.getPipeline(state.pipelines, name, namespace);
}

export function getPipelines(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return pipelineSelectors.getPipelines(state.pipelines, namespace);
}

export function getPipelinesErrorMessage(state) {
  return pipelineSelectors.getPipelinesErrorMessage(state.pipelines);
}

export function isFetchingPipelines(state) {
  return pipelineSelectors.isFetchingPipelines(state.pipelines);
}

export function getPipelineResources(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return pipelineResourcesSelectors.getPipelineResources(
    state.pipelineResources,
    namespace
  );
}

export function getPipelineResource(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return pipelineResourcesSelectors.getPipelineResource(
    state.pipelineResources,
    name,
    namespace
  );
}

export function getPipelineResourcesErrorMessage(state) {
  return pipelineResourcesSelectors.getPipelineResourcesErrorMessage(
    state.pipelineResources
  );
}

export function isFetchingPipelineResources(state) {
  return pipelineResourcesSelectors.isFetchingPipelineResources(
    state.pipelineResources
  );
}

export function getPipelineRuns(
  state,
  { filters, namespace = getSelectedNamespace(state) } = {}
) {
  const runs = pipelineRunsSelectors.getPipelineRuns(
    state.pipelineRuns,
    namespace
  );
  return runs.filter(pipelineRun => {
    return filters.every(filter => {
      const [filterKey, filterValue] = filter.split('=');
      return (
        pipelineRun.metadata.labels &&
        filterKey &&
        filterValue &&
        pipelineRun.metadata.labels[filterKey] === filterValue
      );
    });
  });
}

export function getPipelineRun(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return pipelineRunsSelectors.getPipelineRun(
    state.pipelineRuns,
    name,
    namespace
  );
}

export function getPipelineRunsErrorMessage(state) {
  return pipelineRunsSelectors.getPipelineRunsErrorMessage(state.pipelineRuns);
}

export function isFetchingPipelineRuns(state) {
  return pipelineRunsSelectors.isFetchingPipelineRuns(state.pipelineRuns);
}

export function getTaskRun(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return taskRunsSelectors.getTaskRun(state.taskRuns, name, namespace);
}

export function getTaskRunsByPipelineRunName(
  state,
  pipelineRunName,
  { namespace = getSelectedNamespace(state) } = {}
) {
  const runs = taskRunsSelectors.getTaskRuns(state.taskRuns, namespace);
  return runs.filter(
    taskRun =>
      taskRun.metadata &&
      taskRun.metadata.labels &&
      taskRun.metadata.labels['tekton.dev/pipelineRun'] === pipelineRunName
  );
}

export function getTaskRuns(
  state,
  { filters, namespace = getSelectedNamespace(state) } = {}
) {
  const runs = taskRunsSelectors.getTaskRuns(state.taskRuns, namespace);
  return runs.filter(taskRun => {
    return filters.every(filter => {
      const [filterKey, filterValue] = filter.split('=');
      return (
        taskRun.metadata.labels &&
        filterKey &&
        filterValue &&
        taskRun.metadata.labels[filterKey] === filterValue
      );
    });
  });
}

export function getTaskRunsErrorMessage(state) {
  return taskRunsSelectors.getTaskRunsErrorMessage(state.taskRuns);
}

export function isFetchingTaskRuns(state) {
  return taskRunsSelectors.isFetchingTaskRuns(state.taskRuns);
}

export function getTasks(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return taskSelectors.getTasks(state.tasks, namespace);
}

export function getTask(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return taskSelectors.getTask(state.tasks, name, namespace);
}

export function getTasksErrorMessage(state) {
  return taskSelectors.getTasksErrorMessage(state.tasks);
}

export function isFetchingTasks(state) {
  return taskSelectors.isFetchingTasks(state.tasks);
}

export function getClusterTasks(state) {
  return clusterTaskSelectors.getClusterTasks(state.clusterTasks);
}

export function getClusterTask(state, name) {
  return clusterTaskSelectors.getClusterTask(state.clusterTasks, name);
}

export function getClusterTasksErrorMessage(state) {
  return clusterTaskSelectors.getClusterTasksErrorMessage(state.clusterTasks);
}

export function isFetchingClusterTasks(state) {
  return clusterTaskSelectors.isFetchingClusterTasks(state.clusterTasks);
}

export function getTaskByType(
  state,
  { type, name, namespace = getSelectedNamespace(state) }
) {
  return type === 'clustertasks'
    ? getClusterTask(state, name)
    : getTask(state, { name, namespace });
}

export function getSecrets(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return secretSelectors.getSecrets(state.secrets, namespace);
}

export function getSecretsErrorMessage(state) {
  return secretSelectors.getSecretsErrorMessage(state.secrets);
}

export function getCreateSecretsSuccessMessage(state) {
  return secretSelectors.getCreateSecretsSuccessMessage(state.secrets);
}

export function getDeleteSecretsSuccessMessage(state) {
  return secretSelectors.getDeleteSecretsSuccessMessage(state.secrets);
}

export function isFetchingSecrets(state) {
  return secretSelectors.isFetchingSecrets(state.secrets);
}

export function getLocale(state) {
  return localeSelectors.getLocale(state.locale);
}

export function isWebSocketConnected(state) {
  return notificationSelectors.isWebSocketConnected(state.notifications);
}

export function getEventListeners(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return eventListenerSelectors.getEventListeners(state.eventListeners, namespace);
}

export function getEventListener(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return eventListenerSelectors.getEventListener(state.pipelines, name, namespace);
}

export function getTriggerBindings(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return triggerBindingSelectors.getTriggerBindings(state.triggerBindings, namespace);
}

export function getTriggerBinding(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return eventListenerSelectors.getTriggerBindings(state.triggerBindings, name, namespace);
}

export function getTriggerTemplates(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return triggerTemplateSelectors.getTriggerTemplates(state.triggertemplates, namespace);
}

export function getTriggerTemplate(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return eventListenerSelectors.getTriggerTemplates(state.triggerTemplates, name, namespace);
}
```

carlos-logro commented:
correct but missing `isFetching` and `getErrorMessage`. I was going to include them respectively in the future PR's.

triggerTemplates for example goes:
```
export function getTriggerTemplates(
  state,
  { namespace = getSelectedNamespace(state) } = {}
) {
  return triggerTemplatesSelectors.getTriggerTemplates(
    state.triggerTemplates,
    namespace
  );
}

export function getTriggerTemplate(
  state,
  { name, namespace = getSelectedNamespace(state) }
) {
  return triggerTemplatesSelectors.getTriggerTemplate(
    state.triggerTemplates,
    name,
    namespace
  );
}

export function getTriggerTemplatesErrorMessage(state) {
  return triggerTemplatesSelectors.getTriggerTemplatesErrorMessage(
    state.triggerTemplates
  );
}

export function isFetchingTriggerTemplates(state) {
  return triggerTemplatesSelectors.isFetchingTriggerTemplates(
    state.triggerTemplates
  );
}
```

a-roberts commented:
> correct but missing `isFetching` and `getErrorMessage`. I was going to include them respectively in the future PR's.
> 
> triggerTemplates for example goes:
> 
> ```
> export function getTriggerTemplates(
>   state,
>   { namespace = getSelectedNamespace(state) } = {}
> ) {
>   return triggerTemplatesSelectors.getTriggerTemplates(
>     state.triggerTemplates,
>     namespace
>   );
> }
> 
> export function getTriggerTemplate(
>   state,
>   { name, namespace = getSelectedNamespace(state) }
> ) {
>   return triggerTemplatesSelectors.getTriggerTemplate(
>     state.triggerTemplates,
>     name,
>     namespace
>   );
> }
> 
> export function getTriggerTemplatesErrorMessage(state) {
>   return triggerTemplatesSelectors.getTriggerTemplatesErrorMessage(
>     state.triggerTemplates
>   );
> }
> 
> export function isFetchingTriggerTemplates(state) {
>   return triggerTemplatesSelectors.isFetchingTriggerTemplates(
>     state.triggerTemplates
>   );
> }
> ```

Ok, yeah let's do that then (the follow up PR to add with a combination of our comments above) 😄
